text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,515 @@
+`cargo build -Z check-cfg ...`.
+
+Experience gained during stabilization will determine how this feature is best enabled in the final
+product. Ideally, once the feature is stabilized in `rustc`, the `-Z check-cfg` requirement will
+be dropped from `rustc`. Stabilizing in Cargo may require a stable opt-in flag, however.
+
+## Diagnostics
+
+Conditional checking can report these diagnostics:
+
+* `invalid_cfg_name`: Indicates that a condition name was not in the set of valid names.
+  This diagnostic will only be reported if the command line options enable checking condition names
+  (i.e. there is at least one `--cfg 'names(...)'` option and an invalid condition name is found
+  during compilation.
+
+* `invalid_cfg_value`: Indicates that source code contained a condition value that was invalid.
+  This diagnostic will only be reported if the command line options enable checking condition values
+  for the specified condition name (i.e. there is a least one `--check-cfg 'values(c, ...)'` for
+  a given condition name `c`).
+
+All of the diagnostics defined by this RFC are reported as warnings. They can be upgraded to
+errors or silenced using the usual diagnostics controls.
+
+## Examples
+
+Consider this command line:
+
+```bash
+rustc --check-cfg 'name(feature)' \
+      --check-cfg 'values(feature,"lion","zebra")' \
+      --cfg 'feature="lion"'
+      example.rs
+```
+
+This command line indicates that this crate has two features: `lion` and `zebra`. The `lion`
+feature is enabled, while the `zebra` feature is disabled. Consider compiling this code:
+
+```rust
+// this is valid, and tame_lion() will be compiled
+#[cfg(feature = "lion")]     
+fn tame_lion(lion: Lion) { ... }
+
+// this is valid, and ride_zebra() will NOT be compiled
+#[cfg(feature = "zebra")] 
+fn ride_zebra(zebra: Zebra) { ... }
+
+// this is INVALID, and will cause a compiler error 
+#[cfg(feature = "platypus")] 
+fn poke_platypus() { ... }
+
+// this is INVALID, because 'feechure' is not a known condition name,
+// and will cause a compiler error.
+#[cfg(feechure = "lion")]
+fn tame_lion() { ... }
+```
+
+> Note: The `--check-cfg names(feature)` option is necessary only to enable checking the condition
+> name, as in the last example. `feature` is a well-known (always-valid) condition name, and so it
+> is not necessary to specify it in a `--check-cfg 'names(...)'` option. That option can be
+> shortened to > `--check-cfg names()` in order to enable checking condition names.
+
+## Drawbacks
+
+* Adds complexity, in the form of additional command-line options.
+* As with all lints, correct code may be trigger lints. Developers will need to take time to
+  examine them and see whether they are legitimate or not.
+* To take full advantage of this, build systems (including but not limited to Cargo) must be
+  updated. However, for those systems that are not updated, there is no penalty or drawback,
+  since `--check-cfg` is opt-in.
+

```

Comment:
One other drawback is there might be cases that this lint fails to catch. For example, consider a crate with two files:

```rust
// a.rs

#[cfg(windows)]
mod b;
```

```rust
// b.rs

#[cfg(iwndows)]
pub fn foo() { ... }
```

If `--cfg windows` isn't specified, `mod b` will not be parsed and so the invalid `#[cfg(iwndows)]` condition will not be caught.

Reply:
Hmm, that seems fine though - it will still be caught if you pass `--cfg windows`, and otherwise the cfg is unreachable anyway.

Reply:
My hope is that, at some point, we can have Cargo pass `--check-cfg 'names()'` by default. That would catch this case. That would require some crater runs, mainly to detect places where existing crates are using their own `--cfg foo` options, instead of `--cfg feature="..."`. So I think the sequence is:

1. The feature gets added to `rustc`.
2. We modify `cargo` so that it passes the `--check-cfg 'values(feature, "...")'` option, so we get good coverage for `feature` but not for anything else. It's opt-in behavior, controlled by a field in `Cargo.toml`.
3. We stay in that state for a while, gain experience with it, and deal with any breakage in crates.
4. Once all crates are passing, then _possibly_ we turn this on by default.
5. Another long stabilization period.
6. We modify `cargo` so that it passes `--check-cfg 'names()'`, and then find all of the build scripts that need to be updated. This is an opt-in feature, controlled by a field in `Cargo.toml`.
7. A similar process of updating crates occurs.
8. Eventually, `cargo` turns this behavior on by default.

That's a long process, but I hope that we get most of the value from this at step 2. Maybe `cargo new ...` adds the opt-in flag by default, so that all new crates have opted in.

Reply:
Perhaps I chose a poor example but more generally my point was that because `#[cfg(foo)] mod bla;` is expanded to either `mod bla;` or nothing, in the case where it expands to nothing, the `bla` mod is never loaded by rustc which means any invalid/misspelled `cfg` conditions in that module will not be checked.

Reply:
I've updated the RFC, and added an entry to "Drawbacks" that covers this.